src/ipc-wrapper/server.js

```diff
@@ -0,0 +1,113 @@
+// @ts-check
+import { createServer } from 'rpc-reflector'
+import { MANAGER_CHANNEL_ID, MAPEO_RPC_ID, SubChannel } from './sub-channel.js'
+import { extractMessageEventData } from './utils.js'
+
+/**
+ * @param {import('../mapeo-manager.js').MapeoManager} manager
+ * @param {import('./sub-channel.js').MessagePortLike} messagePort
+ */
+export function createMapeoServer(manager, messagePort) {
+  /** @type {Map<string, { close: () => void }>} */
+  const existingProjectServers = new Map()
+
+  /** @type {Map<string, SubChannel>} */
+  const existingProjectChannels = new Map()
+
+  const mapeoRpcApi = new MapeoRpcApi(manager)
+
+  const managerChannel = new SubChannel(messagePort, MANAGER_CHANNEL_ID)
+  const mapeoRpcChannel = new SubChannel(messagePort, MAPEO_RPC_ID)
+
+  const managerServer = createServer(manager, managerChannel)
+  const mapeoRpcServer = createServer(mapeoRpcApi, mapeoRpcChannel)
+
+  managerChannel.start()
+  mapeoRpcChannel.start()
+
+  messagePort.addEventListener('message', handleMessage)
+
+  return {
+    close() {
+      messagePort.removeEventListener('message', handleMessage)
+
+      for (const [id, server] of existingProjectServers.entries()) {
+        server.close()
+
+        const channel = existingProjectChannels.get(id)
+
+        if (channel) {
+          channel.close()
+          existingProjectChannels.delete(id)
+        }
+
+        existingProjectServers.delete(id)
+      }
+
+      managerServer.close()
+      managerChannel.close()
+      mapeoRpcServer.close()
+    },
+  }
+
+  /**
+   * @param {unknown} payload
+   */
+  async function handleMessage(payload) {
+    const data = extractMessageEventData(payload)
+
+    if (!data || typeof data !== 'object' || !('message' in data)) return
+
+    const id = 'id' in data && typeof data.id === 'string' ? data.id : null
+
+    if (!id || id === MANAGER_CHANNEL_ID || id === MAPEO_RPC_ID) return
+
+    if (existingProjectChannels.has(id)) return
+
+    const projectChannel = new SubChannel(messagePort, id)
+    existingProjectChannels.set(id, projectChannel)
+
+    let project
+    try {
+      project = await manager.getProject(
+        /** @type {import('../types.js').ProjectPublicId} */ (id)
+      )
+    } catch (err) {
+      // TODO: how to respond to client so that method errors?
```

Comment:
for posterity, idea from Gregor is to hook into RPC internals and call `postMessage()` with the encoded RPC message that represents an error that's sent due to method erroring. can be done as follow-up